### PR TITLE
[SPARK-31867][SQL] Fix silent data change for datetime formatting

### PR DESCRIPTION
**Pull request by yaooqinn**
### What changes were proposed in this pull request?

the new datetime formatter introduces silent data change like,

```sql
spark-sql> select from_unixtime(1, 'yyyyyyyyyyy-MM-dd');
NULL
spark-sql> set spark.sql.legacy.timeParserPolicy=legacy;
spark.sql.legacy.timeParserPolicy	legacy
spark-sql> select from_unixtime(1, 'yyyyyyyyyyy-MM-dd');
00000001970-01-01
spark-sql>
```

For patterns that support `SignStyle.EXCEEDS_PAD`, e.g. `y..y`(len >=4), when using the `NumberPrinterParser` to format it

```java
switch (signStyle) {
  case EXCEEDS_PAD:
    if (minWidth < 19 && value >= EXCEED_POINTS[minWidth]) {
      buf.append(decimalStyle.getPositiveSign());
    }
    break;
   
           ....
``` 
the `minWidth` == `len(y..y)`
the `EXCEED_POINTS` is 

```java
/**
         * Array of 10 to the power of n.
         */
        static final long[] EXCEED_POINTS = new long[] {
            0L,
            10L,
            100L,
            1000L,
            10000L,
            100000L,
            1000000L,
            10000000L,
            100000000L,
            1000000000L,
            10000000000L,
        };
```

So when the `len(y..y)` is greater than 10, ` ArrayIndexOutOfBoundsException` will be raised.

 And at the caller side, for `from_unixtime`, the exception will be suppressed and silent data change occurs. for `date_format`, the `ArrayIndexOutOfBoundsException` will continue.

In this PR, for `EXCEPTION` mode, we wrapped exceptions from the new formatter during formatting to `SparkUpgradeException` for end-users to make proper decisions.

### Why are the changes needed?

fix silent data change.

### Does this PR introduce _any_ user-facing change?

Yes, SparkUpgradeException will take place of `null` result when something is wrong during new formatter's formatting phase.

### How was this patch tested?

add unit tests,

the `sql/core/src/test/resources/sql-tests/results/datetime-corrected.sql.out` is added for different behaviors from the caller sides.

### CODE REVIEW

**Comment by yaooqinn** [comment]
cc @cloud-fan @MaxGekk @maropu thanks very much

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeFormatterHelper.scala [hunk 1]
```diff
@@ -118,6 +118,31 @@ trait DateTimeFormatterHelper {
...
+      case _: Throwable => throw e
+    }
+    throw new SparkUpgradeException("3.0", s"Fail to format '$resultCandidate' in the new" +
+      s" formatter. You can set ${SQLConf.LEGACY_TIME_PARSER_POLICY.key} to LEGACY to restore" +
+      s" the behavior before Spark 3.0, or set to CORRECTED and treat it as an invalid" +
+      s" datetime string.", e)
+  }
+
+  private def needConvertToSparkUpgradeException(e: Throwable): Boolean = e match {
+    case _: DateTimeException | _: ArrayIndexOutOfBoundsException
```
Comment (yaooqinn) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeFormatterHelper.scala:141]: Shall we use Nonfatal here?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeFormatterHelper.scala [hunk 2]
```diff
@@ -118,6 +118,31 @@ trait DateTimeFormatterHelper {
...
+    case e if needConvertToSparkUpgradeException(e) =>
+      val resultCandidate = try {
+      legacyFormatFunc(d)
+    } catch {
+      case _: Throwable => throw e
+    }
+    throw new SparkUpgradeException("3.0", s"Fail to format it to '$resultCandidate' in the new" +
+      s" formatter. You can set ${SQLConf.LEGACY_TIME_PARSER_POLICY.key} to LEGACY to restore" +
+      s" the behavior before Spark 3.0, or set to CORRECTED and treat it as an invalid" +
+      s" datetime string.", e)
```
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeFormatterHelper.scala:137]: nit: removes unnecessary `s`s in the head.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeFormatterHelper.scala [hunk 3]
```diff
@@ -105,9 +105,9 @@ trait DateTimeFormatterHelper {
   // between legacy parser and new parser. If new parser fails but legacy parser works, throw a
   // SparkUpgradeException. On the contrary, if the legacy policy set to CORRECTED,
   // DateTimeParseException will address by the caller side.
-  protected def checkDiffResult[T](
+  protected def checkDiffParserResult[T](
       s: String, legacyParseFunc: String => T): PartialFunction[Throwable, T] = {
-    case e: DateTimeException if SQLConf.get.legacyTimeParserPolicy == EXCEPTION =>
+    case e if needConvertToSparkUpgradeException(e) =>
```
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeFormatterHelper.scala:110]: The parser case can get`ArrayIndexOutOfBoundsException`, too?

File: sql/core/src/test/resources/sql-tests/inputs/datetime.sql [hunk 4]
```diff
@@ -160,3 +160,7 @@ select from_json('{"time":"26/October/2015"}', 'time Timestamp', map('timestampF
 select from_json('{"date":"26/October/2015"}', 'date Date', map('dateFormat', 'dd/MMMMM/yyyy'));
 select from_csv('26/October/2015', 'time Timestamp', map('timestampFormat', 'dd/MMMMM/yyyy'));
 select from_csv('26/October/2015', 'date Date', map('dateFormat', 'dd/MMMMM/yyyy'));
+
+-- fix silent data change in date time formatters
+select from_unixtime(1, 'yyyyyyyyyyy-MM-dd');
+select date_format(date '2018-11-17', 'yyyyyyyyyyy-MM-dd');
```
Comment (maropu) [inline sql/core/src/test/resources/sql-tests/inputs/datetime.sql:166]: How about adding a test for the case `4<len(|y....y|)<11`?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeFormatterHelper.scala [hunk 5]
```diff
@@ -118,6 +118,31 @@ trait DateTimeFormatterHelper {
...
+    }
+    throw new SparkUpgradeException("3.0", s"Fail to format it to '$resultCandidate' in the new" +
+      s" formatter. You can set ${SQLConf.LEGACY_TIME_PARSER_POLICY.key} to LEGACY to restore" +
+      s" the behavior before Spark 3.0, or set to CORRECTED and treat it as an invalid" +
+      s" datetime string.", e)
+  }
+
+  private def needConvertToSparkUpgradeException(e: Throwable): Boolean = e match {
+    case _: DateTimeException | _: ArrayIndexOutOfBoundsException
+      if SQLConf.get.legacyTimeParserPolicy == EXCEPTION => true
```
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeFormatterHelper.scala:142]: Could you leave some comments about what's a condition for `ArrayIndexOutOfBoundsException` to be thrown?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeFormatterHelper.scala (hunk 3 above)
Comment (yaooqinn) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeFormatterHelper.scala:110]: There could be such a risk for the parser to get it after roughly checked the JCL.

Comment (yaooqinn) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeFormatterHelper.scala:110]: Maybe RuntimeException is more appropriate

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeFormatterHelper.scala (hunk 1 above)
Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeFormatterHelper.scala:146]: This seems like a Java bug to me to throw `ArrayIndexOutOfBoundsException`.

Shall we fail the formatter creation when there are more than 10 `y` in the pattern string? And document this limitation as well.

**Comment by maropu** [comment]
retest this please

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeFormatterHelper.scala (hunk 1 above)
Comment (yaooqinn) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeFormatterHelper.scala:146]: okay. Do we still need to check the runtime exceptions for the formatting phase?

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeFormatterHelper.scala:146]: Yes, we still need to handle `DateTimeException` as the `DatetimeFormatter.format` method declares to throw `DateTimeException`, although I don't know how.

**Pull request closed without merging**